src/GitHub.TeamFoundation.14/Services/VSGitExt.cs

```diff
@@ -8,91 +9,67 @@
 using GitHub.TeamFoundation.Services;
 using Serilog;
 using Microsoft.VisualStudio.TeamFoundation.Git.Extensibility;
-using Task = System.Threading.Tasks.Task;
 
 namespace GitHub.VisualStudio.Base
 {
     /// <summary>
     /// This service acts as an always available version of <see cref="IGitExt"/>.
     /// </summary>
-    [Export(typeof(IVSGitExt))]
-    [PartCreationPolicy(CreationPolicy.Shared)]
+    /// <remarks>
+    /// Initialization for this service will be done asynchronously and the <see cref="IGitExt" /> service will be
+    /// retrieved using <see cref="GetServiceAsync" />. This means the service can be constructed and subscribed to from a background thread.
+    /// </remarks>
     public class VSGitExt : IVSGitExt
     {
         static readonly ILogger log = LogManager.ForContext<VSGitExt>();
 
-        readonly IGitHubServiceProvider serviceProvider;
-        readonly IVSUIContext context;
+        readonly Func<Type, Task<object>> getServiceAsync;
         readonly ILocalRepositoryModelFactory repositoryFactory;
         readonly object refreshLock = new object();
 
         IGitExt gitService;
         IReadOnlyList<ILocalRepositoryModel> activeRepositories;
 
         [ImportingConstructor]
-        public VSGitExt(IGitHubServiceProvider serviceProvider)
-            : this(serviceProvider, new VSUIContextFactory(), new LocalRepositoryModelFactory())
+        public VSGitExt(Func<Type, Task<object>> getServiceAsync)
+            : this(getServiceAsync, new VSUIContextFactory(), new LocalRepositoryModelFactory())
```

Comment:
Any reason to send a callback instead of sending in `IAsyncServiceProvider` (which provides `GetServiceAsync` here?

Reply:
The issue was that `AsyncPackage` implements 3 interfaces called `IAsyncServiceProvider`, 2 of which exist in the same namespace. Passing a delegate way a way to avoid taking a dependency on `Microsoft.VisualStudio.Shell.Immutable.14.0` and needing to disambiguate using `extern alias` (which works but is a bit messy).

I've pushed a commit that uses the interface from `Microsoft.VisualStudio.Shell.Immutable.14.0`. If we're likely to be using `IAsyncServiceProvider` elsewhere, having the correct assembly to hand is maybe a good thing (even if we need to use `extern alias`).

An alternative approach would be to use the lower level Interop `IAsyncServiceProvider` interface from here:

```
namespace Microsoft.VisualStudio.Shell.Interop
{
    public interface IAsyncServiceProvider
    {
        IVsTask QueryServiceAsync([ComAliasName("OLE.REFGUID")] ref Guid guidService);
    }
}
```

The fact that is takes a `Guid` and returns a `IVsTask` again makes it a bit messy.

My instinct is to go with this last commit. What do you think?